src/core/api/robot.go

```diff
@@ -0,0 +1,197 @@
+// Copyright 2018 Project Harbor Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//    http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package api
+
+import (
+	"fmt"
+	"github.com/goharbor/harbor/src/common/dao"
+	"github.com/goharbor/harbor/src/common/models"
+	"net/http"
+	"strconv"
+)
+
+// User this prefix to distinguish harbor user,
+// The prefix contains a specific character($), so it cannot be registered as a harbor user.
+const robotPrefix = "robot$"
+
+// RobotAPI ...
+type RobotAPI struct {
+	BaseController
+	project *models.Project
+	robot   *models.Robot
+}
+
+// Prepare ...
+func (r *RobotAPI) Prepare() {
+	r.BaseController.Prepare()
+	method := r.Ctx.Request.Method
+
+	if !r.SecurityCtx.IsAuthenticated() {
+		r.HandleUnauthorized()
+		return
+	}
+
+	pid, err := r.GetInt64FromPath(":pid")
+	if err != nil || pid <= 0 {
+		text := "invalid project ID: "
+		if err != nil {
+			text += err.Error()
+		} else {
+			text += fmt.Sprintf("%d", pid)
+		}
+		r.HandleBadRequest(text)
+		return
+	}
+	project, err := r.ProjectMgr.Get(pid)
+	if err != nil {
+		r.ParseAndHandleError(fmt.Sprintf("failed to get project %d", pid), err)
+		return
+	}
+	if project == nil {
+		r.HandleNotFound(fmt.Sprintf("project %d not found", pid))
+		return
+	}
+	r.project = project
+
+	if method == http.MethodPut || method == http.MethodDelete {
+		id, err := r.GetInt64FromPath(":id")
+		if err != nil || id <= 0 {
+			r.HandleBadRequest("invalid robot ID")
```

Comment:
Can we use `Send***Error `of BaseController instead of beego default ones to provide well-format errors?

Reply:
Will update it when to resolve the http error PR.